src/hupper/ipc.py

```diff
@@ -135,8 +136,11 @@ def activate(self, on_recv):
         self.reader_thread.start()
 
     def close(self):
-        close_fd(self.r_fd)
-        close_fd(self.w_fd)
+        self.r_fd, r_fd = -1, self.r_fd
+        self.w_fd, w_fd = -1, self.w_fd
+
+        close_fd(r_fd)
+        close_fd(w_fd)
         self.on_recv = None
```

Comment:
I'm trying to remember why the reader thread wouldn't be joined on here just to ensure proper cleanup.

Reply:
To be clear, wouldn't a `reader_thread.join()` prior to `self.on_recv = None` fix the issue without needing to overwrite the fd values?

Reply:
No, the thread could be executing on_recv when the FD is closed, meaning when it goes around the loop again it tries to read from a stale FD value

Reply:
you also can't join the reader thread because it might not have been created, eg the child_pipe of worker doesn't get "activate()" called on it